### PR TITLE
Introduce common code for Recorders in Plugin Tests

### PR DESCRIPTION
**Pull request by fendor**
Plugin Tests need a recorder. Most plugin test-suite just set it to `mempty`, which, I think, will make it more difficult to debug test-failures.

Some plugin test-suites roll their own plugin initialisation code. This is fine, but makes the logging behaviour non-uniform.

Introduce the `PluginTestDescriptor` type definition.
Utility functions to create `PluginTestDescriptor`'s `mkPluginTestDescriptor`  and `mkPluginTestDescriptor'` which will help migrating plugins to the newer co-log style logger. 

<a href="https://gitpod.io/#https://github.com/haskell/haskell-language-server/pull/3347"></a>

### CODE REVIEW

File: hls-test-utils/src/Test/Hls.hs [hunk 1]
```diff
@@ -134,10 +144,34 @@ goldenWithHaskellDoc plugin title testDataDir path desc ext act =
...
+  -> TestTree
+goldenWithHaskellDocAndRecorder plugin title testDataDir path desc ext act =
+  goldenGitDiff title (testDataDir </> path <.> desc <.> ext)
+  $ runSessionWithServerAndRecorder plugin testDataDir
+  $ TL.encodeUtf8 . TL.fromStrict
+  <$> do
+    doc <- openDoc (path <.> ext) "haskell"
+    void waitForBuildQueue
+    act doc
+    documentContents doc
```
Comment (pepeiborra) [inline hls-test-utils/src/Test/Hls.hs:165]: This completely subsumes `goldenWithHaskellDoc`. Do we need a new function? I would vote for refactoring `goldenWithHaskellDoc` to provide a recorder.

File: hls-test-utils/src/Test/Hls.hs [hunk 2]
```diff
@@ -134,10 +144,34 @@ goldenWithHaskellDoc plugin title testDataDir path desc ext act =
...
 
 runSessionWithServer :: PluginDescriptor IdeState -> FilePath -> Session a -> IO a
 runSessionWithServer plugin = runSessionWithServer' [plugin] def def fullCaps
 
+runSessionWithServerAndRecorder :: Pretty b => (Recorder (WithPriority b) -> PluginDescriptor IdeState) -> FilePath -> Session a -> IO a
+runSessionWithServerAndRecorder pluginF fp act = do
+  recorder <- pluginTestRecorder
+  runSessionWithServer' [pluginF recorder] def def fullCaps fp act
```
Comment (pepeiborra) [inline hls-test-utils/src/Test/Hls.hs:173]: Same about this: I would refactor `runSessionWithServer` to provide a recorder

File: hls-test-utils/src/Test/Hls.hs [hunk 3]
```diff
@@ -211,6 +245,46 @@ keepCurrentDirectory = bracket getCurrentDirectory setCurrentDirectory . const
...
+initialiseTestRecorder :: Pretty a => [String] -> IO (Recorder (WithPriority a), WithPriority (Doc ann) -> IO ())
+initialiseTestRecorder envVars = do
+    docWithPriorityRecorder <- makeDefaultStderrRecorder Nothing Debug
+    -- There are potentially multiple environment variables that enable this logger
+    definedEnvVars <- forM envVars (\var -> fromMaybe "0" <$> lookupEnv var)
+    let logStdErr = any (/= "0") definedEnvVars
+
+        docWithFilteredPriorityRecorder =
+          if logStdErr then mempty
+          else cfilter (\WithPriority{ priority } -> priority >= Debug) docWithPriorityRecorder
```
Comment (pepeiborra) [inline hls-test-utils/src/Test/Hls.hs:282]: maybe I'm confused but I would say that the then-else branches should be the other way around

File: hls-test-utils/src/Test/Hls.hs [hunk 4]
```diff
@@ -211,6 +245,46 @@ keepCurrentDirectory = bracket getCurrentDirectory setCurrentDirectory . const
...
+-- to write all logs to stderr.
+pluginTestRecorder :: Pretty a => IO (Recorder (WithPriority a))
+pluginTestRecorder = do
+  (recorder, _) <- initialiseTestRecorder ["HLS_TEST_PLUGIN_LOG_STDERR"]
+  pure recorder
+
+-- | Generic recorder initialisation for plugins and the HLS server for test-cases.
+--
+-- The created recorder writes to stderr if any of the given environment variables
+-- have been set to a value different to @0@.
```
Comment (pepeiborra) [inline hls-test-utils/src/Test/Hls.hs:267]: Why not write to stderr unconditionally, as the default recorder for HLS does?

File: hls-test-utils/src/Test/Hls.hs (hunk 3 above)
Comment (fendor) [reply hls-test-utils/src/Test/Hls.hs:282]: Yeah you are right, messed up the condition inversion

File: hls-test-utils/src/Test/Hls.hs (hunk 4 above)
Comment (fendor) [reply hls-test-utils/src/Test/Hls.hs:267]: Lots of noise, makes it hard to tell what testcase failed, imo. However, thats also fine by me

Comment (pepeiborra) [reply hls-test-utils/src/Test/Hls.hs:267]: But this is already the status quo, and `lsp-test` silences it anyway, so I'm not sure what the difference is

File: hls-test-utils/src/Test/Hls.hs (hunk 1 above)
Comment (fendor) [reply hls-test-utils/src/Test/Hls.hs:165]: Exactly that was the question, then all plugins that don't use the new recorder infrastructure will have to be updated, too. I don't see a problem with that if you give a thumbs up to that idea.

File: hls-test-utils/src/Test/Hls.hs (hunk 4 above)
Comment (fendor) [reply hls-test-utils/src/Test/Hls.hs:267]: No it is not, the status quo is, we hide the stderr unless `LSP_TEST_LOG_STDERR=1` is given, take a look at this example run with the loggers for HLS and the testsuite writing to stderr vs not writing to stderr: https://paste.tomsmeding.com/0UbJEowE

**Comment by michaelpj** [comment]
> This requires a new sway of test-functions, as long as we have plugins that use the old hslogger functions.

Argh, I forgot we hadn't purged all of that yet >:(

File: hls-test-utils/src/Test/Hls.hs (hunk 3 above)
Comment (michaelpj) [inline hls-test-utils/src/Test/Hls.hs:282]: Might we want an env var for the priority?

**Review by michaelpj — APPROVED** [review]
LGTM!

**Comment by michaelpj** [comment]
I think "always have a recorder" is a good default tbh, so :+1: for refactoring things to make that the main way.

File: hls-test-utils/src/Test/Hls.hs (hunk 3 above)
Comment (fendor) [reply hls-test-utils/src/Test/Hls.hs]: Yeah, but ideally in a follow up PR, this PR just exploded in size :)

File: hls-test-utils/src/Test/Hls.hs (hunk 1 above)
Comment (fendor) [reply hls-test-utils/src/Test/Hls.hs]: Modified as requested

File: hls-test-utils/src/Test/Hls.hs (hunk 3 above)
Comment (fendor) [reply hls-test-utils/src/Test/Hls.hs]: Modified as requested

**Comment by fendor** [comment]
Ready for review, each commit modifies a single plugin so code owners can review exactly the commit they need to.

**Review by konn — APPROVED** [review]
Thank you for great refactoring! LGTM 👍

**Review by ozkutuk — APPROVED** [review]
LGTM!

File: plugins/hls-haddock-comments-plugin/src/Ide/Plugin/HaddockComments.hs [hunk 5]
```diff
@@ -7,7 +7,7 @@
 {-# LANGUAGE RecordWildCards           #-}
 {-# LANGUAGE ViewPatterns              #-}
 
-module Ide.Plugin.HaddockComments (descriptor) where
+module Ide.Plugin.HaddockComments (descriptor, E.Log) where
```
Comment (kokobd) [inline plugins/hls-haddock-comments-plugin/src/Ide/Plugin/HaddockComments.hs:10]: Is this new import item used somewhere?

Comment (fendor) [reply plugins/hls-haddock-comments-plugin/src/Ide/Plugin/HaddockComments.hs:10]: Yes, here: https://github.com/haskell/haskell-language-server/pull/3347/files#diff-52fd7f78c5c4a67b8059732e6f4d8216411b28c7d890cfccc01180c1332740a2R21

**Review by VeryMilkyJoe — APPROVED** [review]

**Review by kokobd — APPROVED** [review]
LGTM!

**Review by georgefst — APPROVED** [review]

**Review by pepeiborra — APPROVED** [review]

**Pull request merged**